app/src/main/java/pl/droidevs/books/di/ViewModelFactory.java

```diff
@@ -2,29 +2,50 @@
 
 import android.arch.lifecycle.ViewModel;
 import android.arch.lifecycle.ViewModelProvider;
-import android.support.annotation.NonNull;
 
-import javax.inject.Inject;
+import java.util.Map;
 
-import pl.droidevs.books.library.LibraryViewModel;
+import javax.inject.Inject;
+import javax.inject.Provider;
 
 public class ViewModelFactory implements ViewModelProvider.Factory {
 
-    private LibraryViewModel viewModel;
+    private final Map<Class<? extends ViewModel>, Provider<ViewModel>> creators;
 
     @Inject
-    public ViewModelFactory(LibraryViewModel libraryViewModel) {
-        this.viewModel = libraryViewModel;
+    public ViewModelFactory(Map<Class<? extends ViewModel>, Provider<ViewModel>> creators) {
+        this.creators = creators;
     }
 
-    @NonNull
+    @SuppressWarnings("unchecked")
     @Override
-    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
+    public <T extends ViewModel> T create(Class<T> modelClass) {
+        Provider<? extends ViewModel> creator = this.creators.get(modelClass);
+
+        if (creator == null) {
+            creator = getCreatorFromCreatorSet(modelClass);
+        }
+
+        if (creator == null) {
```

Comment:
Czy nie lepiej będzie rzucić wyjątkiem bezpośrednio z metody getCreatorFromCreatorSet? Wtedy to sprawdzenie nie będzie potrzebne.